libraries/pushproviders/firebase/src/main/kotlin/io/element/android/libraries/pushproviders/firebase/FirebaseTokenDeleter.kt

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright 2023, 2024 New Vector Ltd.
+ *
+ * SPDX-License-Identifier: AGPL-3.0-only
+ * Please see LICENSE in the repository root for full details.
+ */
+
+package io.element.android.libraries.pushproviders.firebase
+
+import com.google.firebase.messaging.FirebaseMessaging
+import com.squareup.anvil.annotations.ContributesBinding
+import io.element.android.libraries.di.AppScope
+import timber.log.Timber
+import javax.inject.Inject
+import kotlin.coroutines.resume
+import kotlin.coroutines.resumeWithException
+import kotlin.coroutines.suspendCoroutine
+
+interface FirebaseTokenDeleter {
+    suspend fun delete()
+}
+
+/**
+ * This class deletes the current Firebase token.
+ */
+@ContributesBinding(AppScope::class)
+class DefaultFirebaseTokenDeleter @Inject constructor(
+    private val isPlayServiceAvailable: IsPlayServiceAvailable,
+) : FirebaseTokenDeleter {
+    override suspend fun delete() {
+        suspendCoroutine { continuation ->
+            // 'app should always check the device for a compatible Google Play services APK before accessing Google Play services features'
+            if (isPlayServiceAvailable.isAvailable()) {
```

Comment:
Not a needed change, but an early return here with `!isPlayServiceAvailable.isAvailable()` might be more readable. Also, for the exception, maybe a `throw Exception(...).also(Timber::e)` might work?

Reply:
I will change. I am also wondering why we throw the exception here and we do not invoke `continuation.resumeWithException(e)` 🤔

Reply:
12f5839eefa066dce490a40775bbdaf01f902076